libres/lib/enkf/state_map.cpp

```diff
@@ -136,140 +84,141 @@ bool state_map_legal_transition(realisation_state_enum state1,
         return false;
 }
 
-static void state_map_assert_writable(const state_map_type *map) {
-    if (map->read_only)
+void StateMap::m_assert_writable() const {
+    if (m_read_only)
         util_abort("%s: tried to modify read_only state_map - aborting \n",
                    __func__);
 }
 
-static void state_map_iset__(state_map_type *map, int index,
-                             realisation_state_enum new_state) {
-    realisation_state_enum current_state =
-        (realisation_state_enum)int_vector_safe_iget(map->state, index);
+void StateMap::set(int index, realisation_state_enum new_state) {
+    m_assert_writable();
+    std::lock_guard lock{m_mutex};
+
+    if (index < 0)
+        index += m_state.size();
+    if (index < 0)
+        throw std::out_of_range(
+            fmt::format("index out of range: {} < 0", index));
+    if (index >= m_state.size())
+        m_state.resize(index + 1, STATE_UNDEFINED);
```

Comment:
I know this is the old behavior, but it might makes sense to at least write an issue with future changes, this being one of the key ones, it is quite confusing behavior.

Reply:
https://github.com/equinor/ert/issues/3493